components/org.wso2.carbon.uis/src/main/java/org/wso2/carbon/uis/internal/io/deployment/ArtifactAppDeployer.java

```diff
@@ -27,125 +27,170 @@
 
     /**
-     * Creates a new app deployer that locates apps from {@code <CARBON_HOME>/deployment/reactapps} directory.
+     * Creates a new app deployer.
      */
     public ArtifactAppDeployer() {
-        this(Paths.get(System.getProperty("carbon.home", "."), "deployment", "reactapps"));
-    }
-
-    /**
-     * Creates a new app deployer that locates apps from the given directory.
-     *
-     * @param appsRepository app repository directory
-     */
-    public ArtifactAppDeployer(Path appsRepository) {
-        this.appsRepository = appsRepository;
+        this.artifactType = new ArtifactType<>(ARTIFACT_TYPE);
+        URL deploymentLocationUrl = null;
+        try {
+            deploymentLocationUrl = new URL(DEPLOYMENT_LOCATION);
+        } catch (MalformedURLException e) {
+            LOGGER.error("Invalid URL '{}' as app deployment location.", DEPLOYMENT_LOCATION);
+        }
+        this.deploymentLocation = deploymentLocationUrl;
+        this.deployedApps = new ConcurrentHashMap<>();
     }
 
     @Reference(name = "deploymentListener",
                service = AppDeploymentEventListener.class,
-               cardinality = ReferenceCardinality.AT_LEAST_ONE,
+               cardinality = ReferenceCardinality.MANDATORY,
                policy = ReferencePolicy.DYNAMIC,
                unbind = "unregisterListener")
     protected void registerListener(AppDeploymentEventListener listener) {
-        this.eventListeners.add(listener);
+        this.appDeploymentEventListener = listener;
         LOGGER.debug("An instance of class '{}' registered as an app deployment listener.",
                      listener.getClass().getName());
     }
 
     protected void unregisterListener(AppDeploymentEventListener listener) {
-        this.eventListeners.remove(listener);
+        this.appDeploymentEventListener = null;
         LOGGER.debug("An instance of class '{}' unregistered as an app deployment listener.",
                      listener.getClass().getName());
     }
 
     @Activate
     protected void activate(BundleContext bundleContext) {
-        start();
         LOGGER.debug("Carbon UI server app deployer activated.");
     }
 
     @Deactivate
     protected void deactivate(BundleContext bundleContext) {
-        stop();
+        try {
+            for (String artifactKey : deployedApps.keySet()) {
+                undeploy(artifactKey);
+            }
+        } catch (CarbonDeploymentException e) {
+            throw new UISRuntimeException("An error occurred during deactivation of app deployer.", e);
+        }
         LOGGER.debug("Carbon UI server app deployer deactivated.");
     }
 
-    private void start() {
-        if (!Files.exists(appsRepository)) {
-            LOGGER.debug("Web apps repository '{}' does not exists.", appsRepository.toString());
-            return;
-        }
+    @Override
+    public void init() {
+        LOGGER.debug("Carbon UI server app deployer initialized.");
```

Comment:
Better check Logger.isDebugEnabled. Do it at all debug logs.

Reply:
Here: 
SLF4J internally checks whether DEBUG level is enabled or not (see [Slf4jLogger class](https://github.com/ops4j/org.ops4j.pax.logging/blob/logging-1.10.0/pax-logging-api/src/main/java/org/ops4j/pax/logging/slf4j/Slf4jLogger.java#L272)). So we don't need to check it.

In other places:
Since we are using parameterized logs with SLF4J, no need for `if(LOGGER.isDebugEnabled())` check. See [SLF4J FAQ on logging performance](https://www.slf4j.org/faq.html#logging_performance).

There are some places that you need to have the `if(LOGGER.isDebugEnabled())`, however such situation hasn't occurred in Carbon UI Server. (refer [SO answer](https://stackoverflow.com/a/12953090)).

Reply:
Ack. thanks.